File: servicetaskmanager/src/main/java/com/example/alexeyglushkov/authtaskmanager/ServiceTaskImp.java
package com.example.alexeyglushkov.authtaskmanager;

import com.example.alexeyglushkov.authorization.requestbuilder.HttpUrlConnectionBuilder;
import com.example.alexeyglushkov.taskmanager.task.SimpleTask;
import com.example.alexeyglushkov.taskmanager.task.Task;
import com.example.alexeyglushkov.taskmanager.task.TaskBase;
import com.example.alexeyglushkov.taskmanager.task.TaskImpl;

/**
 * Created by alexeyglushkov on 17.07.16.
 */
public abstract class ServiceTaskImp<T> extends BaseServiceTask<T> {
    private TaskBase task;

    public ServiceTaskImp() {
        setTask(new TaskImpl() {
            @Override
            public void startTask() {
                onStart();

                getPrivate().handleTaskCompletion();
            }
        });
    }

    public abstract void onStart();

    protected void setResult(T result) {
        task.getPrivate().setTaskResult(result);
    }

    protected void setError(Error err) {
        task.getPrivate().setTaskError(err);
    }
}

File: servicetaskmanager/src/main/java/com/example/alexeyglushkov/authtaskmanager/ServiceTaskImp.kt
package com.example.alexeyglushkov.authtaskmanager

import com.example.alexeyglushkov.taskmanager.task.TaskImpl
import java.lang.Exception

/**
 * Created by alexeyglushkov on 17.07.16.
 */
abstract class ServiceTaskImp<T> : BaseServiceTask<T>() {
    init {
        task = object : TaskImpl() {
            override suspend fun startTask() {
                onStart()
            }
        }
    }

    abstract fun onStart()

    protected fun setResult(result: T) {
        task.private.taskResult = result
    }

    protected fun setError(err: Exception) {
        task.private.taskError = err
    }
}
